### PR TITLE
Add save_with_initial_vote method to post inside a transaction.

### PR DESCRIPTION
**Pull request by UJKhokhar**

### CODE REVIEW

**Comment by chikamichi** [comment]
Hey,

what we want to do here, is dealing with a two-step save operation. Starting from:

``` ruby
if @post.save
  @post.create_vote
  flash[:notice] = "Post was saved."
  redirect_to [@topic, @post]
else
  # …
end
```

we'd like to perform the `@post.create_vote` not *after* the `post` has been save (`if @post.save`), but *at the same time*. Currently, if the `post` is successfully saved but the vote creation fails somehow, it's too late to catch that error and change our mind: we'd still notice the user everything went fine and redirect to the post.

The idea is to create a new method, different from the regular `save` (which should remain "pure", ie. no side effects such as creating a default vote value). That method would both save and create a vote, and it should error if any of those two operations were to fail.

Basically, it could be:

``` ruby
def save_with_initial_vote
  save
  create_vote
end
```

but it wouldn't behave as an *atomic* operation that can be canceled as a whole if any of its "sub-operation" was to fail ([from the documentation](http://api.rubyonrails.org/classes/ActiveRecord/Transactions/ClassMethods.html): "Transactions are protective blocks where SQL statements are only permanent if they can all succeed as one atomic action").

In order to transform that into something *atomic*, we need to turn it into a transaction:

``` ruby
def save_with_initial_vote
  transaction do
    save
    create_vote
  end
end
```

In the context of a model file, such as `app/models/post.rb`, `transaction` is a shortcut for `ActiveRecord::Base.transaction`.

**Comment by chikamichi** [comment]
Before you modify this branch ("save-with-initial-vote-transaction") by pushing a new commit, I'd like to show you how you can edit it *in place* (ie. updating the pull request). Let's do that during our next meeting :)

**Comment by chikamichi** [comment]
So, we saw during our last meeting how to edit the last recorded commit, using `git commit --amend`.

In order to update that PR, you would:

``` sh
git checkout save-with-initial-vote-transaction
# edit app/models/post.rb
git add app/models/post.rb
git commit --amend
# Because we've just changed the history, we need to "force" push to override what's on GitHub already
git push -f
```

And then, you can merge the PR here, and update your local `master` branch:

``` sh
git checkout master
git pull
# Let's delete our now merged and useless topic branch
git branch -d save-with-initial-vote-transaction
# If you want, you can also delete the remote topic branch, using a weird syntax
git push origin --delete save-with-initial-vote-transaction
```
